### PR TITLE
Feature/wallet connect develop

### PR DESCRIPTION
**Pull request by bnsports**

### CODE REVIEW

**Review by Viscount88 — DISMISSED** [review]

File: fearless/Modules/WalletConnectActiveSessions/WalletConnectActiveSessionsViewLayout.swift [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            make.top.equalTo(tableView.snp.bottom).offset(UIConstants.bigOffset)
+            make.leading.trailing.equalToSuperview().inset(UIConstants.bigOffset)
+            make.bottom.equalTo(safeAreaLayoutGuide)
+            make.height.equalTo(UIConstants.actionHeight)
+        }
+    }
+
+    private func applyLocalization() {
+        navigationBar.setTitle("Connections")
+        searchView.textField.placeholder = "Search by connection"
```
Comment (bnsports) [inline fearless/Modules/WalletConnectActiveSessions/WalletConnectActiveSessionsViewLayout.swift:79]: localization

File: fearless/Modules/WalletConnectConfirmation/WalletConnectConfirmationViewLayout.swift [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        view.valueLabel.font = .h5Title
+        view.valueLabel.textColor = R.color.colorWhite()
+        view.valueLabel.lineBreakMode = .byTruncatingMiddle
+        view.borderView.isHidden = true
+        view.equalsLabelsWidth = true
+        return view
+    }
+
+    private func applyLocalization() {
+        navigationBar.setTitle("Preview")
```
Comment (bnsports) [inline fearless/Modules/WalletConnectConfirmation/WalletConnectConfirmationViewLayout.swift:192]: localization

File: fearless/Modules/WalletConnectProposal/Model/WalletConnectProposalViewModelFactory.swift [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+            chains: chains,
+            cellTitle: "Review required permissions"
+        ) else {
+            throw JSONRPCError.unauthorizedChain
+        }
+
+        let optionalExpandable = try? createProposalPermissionsViewModel(
+            from: proposal.optionalNamespaces,
+            chains: chains,
+            cellTitle: "Review optional permissions"
```
Comment (bnsports) [inline fearless/Modules/WalletConnectProposal/Model/WalletConnectProposalViewModelFactory.swift:71]: localization

File: fearless/Modules/WalletConnectProposal/Model/WalletConnectProposalViewModelFactory.swift [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+        chains: [ChainModel],
+        wallets: [MetaAccountModel],
+        locale _: Locale
+    ) throws -> WalletConnectProposalViewModel {
+        let dApp = createDAppViewModel(from: proposal)
+
+        guard let requiredNetwroks = try createNetworksViewModel(
+            from: proposal.requiredNamespaces,
+            chains: chains,
+            title: "Required networks"
```
Comment (bnsports) [inline fearless/Modules/WalletConnectProposal/Model/WalletConnectProposalViewModelFactory.swift:49]: localization

File: fearless/Modules/WalletConnectProposal/Model/WalletConnectProposalViewModelFactory.swift [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+        ) else {
+            throw JSONRPCError.unauthorizedChain
+        }
+
+        let optionalNetworks = try? createNetworksViewModel(
+            from: proposal.optionalNamespaces,
+            chains: chains,
+            title: "Optional networks"
```
Comment (bnsports) [inline fearless/Modules/WalletConnectProposal/Model/WalletConnectProposalViewModelFactory.swift:57]: localization

File: fearless/Modules/WalletConnectProposal/Model/WalletConnectProposalViewModelFactory.swift [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+        )
+
+        guard let requiredExpandable = try createProposalPermissionsViewModel(
+            from: proposal.requiredNamespaces,
+            chains: chains,
+            cellTitle: "Review required permissions"
```
Comment (bnsports) [inline fearless/Modules/WalletConnectProposal/Model/WalletConnectProposalViewModelFactory.swift:63]: localization

File: fearless/Modules/WalletConnectProposal/WalletConnectProposalExpandableTableCell.swift [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        expandableContentStack.addArrangedSubview(eventsLabel)
+
+        expandableContentStack.setCustomSpacing(UIConstants.defaultOffset, after: chainNameLabel)
+        expandableContentStack.setCustomSpacing(UIConstants.minimalOffset, after: methodsTitleLabel)
+        expandableContentStack.setCustomSpacing(UIConstants.bigOffset, after: methodsLabel)
+        expandableContentStack.setCustomSpacing(UIConstants.minimalOffset, after: eventsTitleLabel)
+    }
+
+    private func applyLocalization() {
+        methodsTitleLabel.text = "Methods"
```
Comment (bnsports) [inline fearless/Modules/WalletConnectProposal/WalletConnectProposalExpandableTableCell.swift:168]: localization

File: fearless/Modules/WalletConnectProposal/WalletConnectProposalPresenter.swift [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+        view?.didReceive(viewModel: updatedViewModel)
+        viewModel = updatedViewModel
+    }
+
+    private func submitReject() {
+        Task {
+            do {
+                guard let proposal = status.proposal else { return }
+                try await interactor.submit(proposalDecision: .reject(proposal: proposal))
+                await showAllDone(description: "Rejected")
```
Comment (bnsports) [inline fearless/Modules/WalletConnectProposal/WalletConnectProposalPresenter.swift:123]: localization

File: fearless/Modules/WalletConnectProposal/WalletConnectProposalPresenter.swift [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+                }
+                let namespaces = try walletConnectModelFactory.createSessionNamespaces(
+                    from: proposal,
+                    wallets: selectedWallets,
+                    chains: chains,
+                    optionalChainIds: optionalChainsIds
+                )
+                try await interactor.submit(proposalDecision: .approve(proposal: proposal, namespaces: namespaces))
+                let dApp = proposal.proposer.name
+                await showAllDone(description: "Сonnection from \(dApp) with ethers has been successfully completed")
```
Comment (bnsports) [inline fearless/Modules/WalletConnectProposal/WalletConnectProposalPresenter.swift:146]: localization

File: fearless/Modules/WalletConnectProposal/WalletConnectProposalPresenter.swift [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+                handle(error: error)
+            }
+        }
+    }
+
+    private func submitDisconnect(topic: String, name: String) {
+        Task {
+            do {
+                try await interactor.submitDisconnect(topic: topic)
+                await showAllDone(description: "Disconnection from React \(name) ethers has been successfully completed")
```
Comment (bnsports) [inline fearless/Modules/WalletConnectProposal/WalletConnectProposalPresenter.swift:158]: localization

File: fearless/Modules/WalletConnectProposal/WalletConnectProposalPresenter.swift [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+            moduleOutput: self,
+            view: view
+        )
+    }
+
+    private func showAllDone(description: String) async {
+        await MainActor.run {
+            view?.didStopLoading()
+            router.showAllDone(
+                title: "All Done",
```
Comment (bnsports) [inline fearless/Modules/WalletConnectProposal/WalletConnectProposalPresenter.swift:198]: localization

File: fearless/Modules/WalletConnectProposal/WalletConnectProposalViewLayout.swift [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+            make.leading.trailing.equalToSuperview().inset(UIConstants.bigOffset)
+            make.bottom.equalTo(safeAreaLayoutGuide)
+        }
+    }
+
+    private func applyLocalization() {
+        switch status {
+        case .proposal:
+            navigationBar.setTitle("Connection details")
+            mainActionButton.imageWithTitleView?.title = "Approve"
```
Comment (bnsports) [inline fearless/Modules/WalletConnectProposal/WalletConnectProposalViewLayout.swift:110]: localization

File: fearless/Modules/WalletConnectProposal/WalletConnectProposalViewLayout.swift [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+            rejectButton.imageWithTitleView?.title = "Reject"
+        case .active:
+            navigationBar.setTitle("Connect details")
```
Comment (bnsports) [inline fearless/Modules/WalletConnectProposal/WalletConnectProposalViewLayout.swift:113]: localization

File: fearless/Modules/WalletConnectSession/WalletConnectSessionViewLayout.swift [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+        stack.distribution = .fillProportionally
+
+        stack.addArrangedSubview(iconView)
+        stack.addArrangedSubview(label)
+
+        return stack
+    }
+
+    private func applyLocalization() {
+        messageView.title = "Message"
```
Comment (bnsports) [inline fearless/Modules/WalletConnectSession/WalletConnectSessionViewLayout.swift:208]: localization

File: fearless/Modules/WalletConnectSession/WalletConnectSessionViewModelFactory.swift [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+            walletName: wallet.name,
+            fiatBalance: totalFiatValue,
+            dayChange: dayChange
+        )
+
+        return viewModel
+    }
+
+    private func createWarning(locale _: Locale) -> NSAttributedString {
+        let warning = NSMutableAttributedString(string: "WARNING: Signing this message can have dangerous side effect. Only sign message from sites you fully trust with your entire account.")
```
Comment (bnsports) [inline fearless/Modules/WalletConnectSession/WalletConnectSessionViewModelFactory.swift:139]: localization

File: fearless/Modules/WalletConnectSession/WalletConnectSessionViewModelFactory.swift [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+
+        warning.addAttribute(
+            NSAttributedString.Key.foregroundColor,
+            value: R.color.colorOrange()!.cgColor,
+            range: NSRange(
+                location: 0,
+                length: "WARNING:".count
```
Comment (bnsports) [inline fearless/Modules/WalletConnectSession/WalletConnectSessionViewModelFactory.swift:146]: localization

File: fearless/Modules/WalletConnectSession/WalletConnectSessionViewModelFactory.swift [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+            )
+        )
+
+        warning.addAttribute(
+            NSAttributedString.Key.font,
+            value: UIFont.h6Title,
+            range: NSRange(
+                location: 0,
+                length: "WARNING:".count
```
Comment (bnsports) [inline fearless/Modules/WalletConnectSession/WalletConnectSessionViewModelFactory.swift:155]: localization

File: fearless/Modules/WalletConnectSession/WalletConnectSessionViewModelFactory.swift [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+            )
+        )
+
+        warning.addAttribute(
+            NSAttributedString.Key.font,
+            value: UIFont.p2Paragraph,
+            range: NSRange(
+                location: "WARNING:".count,
+                length: warning.string.count - "WARNING:".count
```
Comment (bnsports) [inline fearless/Modules/WalletConnectSession/WalletConnectSessionViewModelFactory.swift:164]: localization

File: fearless/Modules/WalletConnectSession/WalletConnectSessionViewModelFactory.swift [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+            )
+        )
+
+        warning.addAttribute(
+            NSAttributedString.Key.foregroundColor,
+            value: R.color.colorStrokeGray()!.cgColor,
+            range: NSRange(
+                location: "WARNING:".count,
+                length: warning.string.count - "WARNING:".count
```
Comment (bnsports) [inline fearless/Modules/WalletConnectSession/WalletConnectSessionViewModelFactory.swift:173]: localization

File: fearless/Modules/WalletMainContainer/WalletMainContainerRouter.swift [hunk 20]
```diff
@@ -1,6 +1,7 @@
 import Foundation
 import UIKit
 import SSFModels
+// import WalletConnectSwiftV2
```
Comment (bnsports) [inline fearless/Modules/WalletMainContainer/WalletMainContainerRouter.swift:4]: commented

File: fearless/ja.lproj/Localizable.strings [hunk 21]
```diff
@@ -1057,4 +1057,6 @@
 "your.validators.change.validators.title" = "バリデーターの変更";
 "your.validators.stop.nominating.title" = "ノミネートを中止";
 "your.validators.validator.total.stake" = "総ステーク：%@";
-"сurrencies.stub.text" = "通貨";
\ No newline at end of file
+"сurrencies.stub.text" = "通貨";
+"error.unsupported.asset" = "You're trying to make a transfer of the asset which isn't currently supported on the App. Please choose another asset or request another QR code.";
```
Comment (bnsports) [inline fearless/ja.lproj/Localizable.strings:1061]: missed localization

File: fearless/ru.lproj/Localizable.strings [hunk 22]
```diff
@@ -1161,4 +1161,6 @@ Euro cash";
 "verify.phone.number.code.input.field.label" = "SMS-код";
 "verify.phone.number.description" = "Введите SMS-код, который был отправлен\n%@";
 "verify.phone.number.send.code" = "ОТПРАВИТЬ КОД";
-"verify.phone.number.title" = "Проверьте свой номер телефона";
\ No newline at end of file
+"verify.phone.number.title" = "Проверьте свой номер телефона";
+"error.unsupported.asset" = "You're trying to make a transfer of the asset which isn't currently supported on the App. Please choose another asset or request another QR code.";
```
Comment (bnsports) [inline fearless/ru.lproj/Localizable.strings:1165]: missed localization

File: fearless/zh-Hans.lproj/Localizable.strings [hunk 23]
```diff
@@ -1158,4 +1158,6 @@
 "verify.phone.number.code.input.field.label" = "短信验证码";
 "verify.phone.number.description" = "输入已发送到\n%@的短信验证码";
 "verify.phone.number.send.code" = "发送验证码";
-"verify.phone.number.title" = "验证您的电话号码";
\ No newline at end of file
+"verify.phone.number.title" = "验证您的电话号码";
+"error.unsupported.asset" = "You're trying to make a transfer of the asset which isn't currently supported on the App. Please choose another asset or request another QR code.";
```
Comment (bnsports) [inline fearless/zh-Hans.lproj/Localizable.strings:1162]: missed localization

File: fearless/ApplicationLayer/Services/Model/JSONRPCError+Fearless.swift [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+    )
+
+    static let unsupportedAccounts = JSONRPCError(
+        code: 5103,
+        message: "Unsupported or empty accounts for namespace"
+    )
+}
+
+extension JSONRPCError: ErrorContentConvertible {
+    func toErrorContent(for _: Locale?) -> ErrorContent {
```
[the remark this answers is not in the document]
Comment (DRadmir) [reply fearless/ApplicationLayer/Services/Model/JSONRPCError+Fearless.swift]: protocol method

File: fearless/Common/Model/ChainRegistry/ChainModel.swift [hunk 25]
```diff
@@ -19,3 +19,26 @@ extension ChainModel: Identifiable {
...
     }
 }
+
+// MARK: - Wallet connect
+
+extension ChainModel {
+    func match(_ caip2ChainId: Caip2ChainId) -> Bool {
+        switch chainBaseType {
+        case .substrate:
+            let namespace = "polkadot"
```
[the remark this answers is not in the document]
Comment (DRadmir) [reply fearless/Common/Model/ChainRegistry/ChainModel.swift:29]: expected. Polkadot like ecosystem

File: fearless/Modules/Send/SendPresenter.swift [hunk 26]
```diff
@@ -80,686 +98,1002 @@ final class SendPresenter {
...
 
-    func didFinishWith(address: String) {
-        searchTextDidChanged(address)
+            interactor.updateSubscriptions(for: qrChainAsset)
+            self.bokoloCashId = bokoloCashId
+            await MainActor.run { [isUserInteractiveAmount] in
+                view?.didReceive(viewModel: viewModel)
+                provideInputViewModel()
+                let networkViewModel = SelectNetworkViewModel(
+                    chainName: "Bokolo cash",
```
[the remark this answers is not in the document]
Comment (DRadmir) [reply fearless/Modules/Send/SendPresenter.swift:672]: Bokolo cash this is just label, not chain. Not localisable label

**Review by DRadmir — APPROVED** [review]

**Pull request merged**